### DISCUSSION TITLE
Page Break Issue with Long Lists in Callout Boxes for PDF Rendering

### DISCUSSION THREAD

**Discussion by produnis** [quarto-dev/quarto-cli#11212] [Q&A] [answered]
### Description

I’m encountering an issue with long lists inside a callout box when rendering to PDF.

You can find a minimal example here: https://github.com/produnis/minimal_examples/tree/longlist.

The problem is that the content does not break across pages; instead, any content that reaches the page end is not displayed in the PDF output.

Is there a solution or workaround for this issue? I am rendering the document to HTML, EPUB, and PDF.

Tested with quarto 1.5.57 and 1.6.30

**Comment by mcanouil** [comment]
I'm not sure callout is the thing you want when it cannot be contained in one single page or when it's long.
I don't think it was thought initially to contain that much.

This being said, PDF (LaTeX) is very tricky regarding this particular type of problems. I'm not sure there is a fix.
One thing you could try is to investigate, on your end, the LaTeX code using `keep-tex: true` to tweak the output and compile it with `xelatex`, `pdflates`, etc.


Note that the content is displayed, it simply overflow the page.

**Reply by mcanouil** [reply]
I was going to suggest Typst instead of PDF/LaTeX, but the issue remains nearly the same.
<img width="421" alt="image" src="https://github.com/user-attachments/assets/69d1cfb2-8058-41fc-8e41-c59c3a1d438b">

May I ask, why use a callout instead of a regular section? In my opinion, it looks weird to have a full page callout. Even weirder, if the content spilled over the second page.

**Reply by produnis** [reply]
Hi there, and thx for your efforts...
my book is about exercises in R and data.table. While there are "normal" sections giving some basic information about how to use data.table, the exercises are all in blue callout-boxes, to give a visual hint that they belong together. All answers are in green callout-boxes. this looks just nice, have a live example here: https://www.produnis.de/tabletrainer  (work still in progress)

**Reply by mcanouil** [reply]
I see thanks for the added context.

May I suggest an alternative which would also allow those "section callout" to appear in table of contents?

Colour the section titles and add icons to it (both extensions work for HTML/PDF):
- https://github.com/mcanouil/quarto-highlight-text
- https://github.com/quarto-ext/fontawesome

**Reply by produnis** [reply]
thx, interesting, I will dive into it and check...

**Reply by produnis** [reply]
here is my dirty hack:

I manually looked how many items would fit on the page without pagebreak. I split the callout-box at this number of items. Also, I wrapped a div called `.custom-counter` around the second callout-box

So it looks like this:
```
::: {.callout-note appearance="simple"}

a) first
b) second
c) ----
d) till it reaches end of page
:::

\newpage

::::{.custom-counter}
::: {.callout-note appearance="minimal"}
I changed to "minimal" here
e) continue
f) with
g) items
:::
::::
```
In LaTex, you cant see anything of this "hack". It just looks nice, and as I use a) b) c), the numbering is correct within the second callout-box.

In HTML (and epub), you can see the split of the callout-box, but as I use "minimal" for the second box, it is acceptable.
Also, to make the counter catch the correct numbering in the second box, I added to my CSS-file:

```
.custom-counter ol {
  list-style-type: none;
  counter-reset: list-counter 13;

.custom-counter ol > li {
  counter-increment: list-counter;
}

.custom-counter ol > li::before {
  content: counter(list-counter, lower-alpha) ". ";
}
```

The `counter-reset: list-counter 13;` makes the list start at the right "number" (which is n) in my case).

So, I can live with that...

**Reply by mcanouil** [reply]
You could probably instead use a Lua filter only for LaTeX, where you loop over the items in callout and insert page break.

**Reply by produnis** [reply]
50,- Euro for the nice guy creating this Lua filter, via paypal or gift card :-)

**Reply by produnis** [reply]
> I was going to suggest Typst instead of PDF/LaTeX, but the issue remains nearly the same. <img alt="image" width="421" src="https://private-user-images.githubusercontent.com/8896044/380521001-69d1cfb2-8058-41fc-8e41-c59c3a1d438b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE1MzcxNjcsIm5iZiI6MTczMTUzNjg2NywicGF0aCI6Ii84ODk2MDQ0LzM4MDUyMTAwMS02OWQxY2ZiMi04MDU4LTQxZmMtOGU0MS1jNTljM2ExZDQzOGIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTExMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMTNUMjIyNzQ3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YTkwOTVmMzM0MDgyNjk1ZmUwYmI3ZjU3NTlhZDQyMmU5MGZhM2M2NjdkMzA4MDU5OWY4NWRiYWRhYzAxNDA2MSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.712uJjxYKm0_xAY7Wf3z699LrGV1oOsWdkutc6shZ2Q">


I tested around with typst these days, and I found that callout boxes are not breakable at all in typst format. Is there a specific reason why?

**Reply by mcanouil** [reply]
Likely because floats support is at its premise.
The next Typst (to be embedded in Quarto will improve floats in general).
Typst is pretty new and some of the features also depends on Pandoc supporting them first.

**Comment by produnis** [comment]
While stuck in German "Schienenersatzverkehr" (possibly the worst word you can hear), I experimented with `keep-tex: true` and manual PDF compilation.

I found that the `minipage` environment in callout boxes was preventing page breaks within long lists. Each callout box uses two `minipage` environments—one for the header and another for the body. By removing the second `minipage`...

```
\begin{minipage}[t]{\textwidth - 5.5mm}

(my items, don’t remove these)

\end{minipage}
```

...everything renders perfectly.

The ideal solution would be to create a Lua filter that removes both `\begin{minipage}[t]{\textwidth - 5.5mm}` and the corresponding `\end{minipage}`.

Tomorrow, I'll have Schienenersatzverkehr again, and maybe I'll get this filter done on my own..
:)
(but any help is highly appreciated)

**Reply by produnis** [reply]
I seem to have the wrong pattern 

```
            local start_pattern = "\\begin%{minipage%}%[t%]%{\\textwidth - 5.5mm%}"
            local end_pattern = "\\end%{minipage%}"
            raw_el.text = raw_el.text:gsub(start_pattern, "")
            raw_el.text = raw_el.text:gsub(end_pattern, "")
```
this does not remove the minipage. Am I escaping the wrong way?

(see full script/exampe here https://github.com/produnis/minimal_examples/tree/longlist )

**Reply by produnis** [reply]
I think I might have an idea why my code is not working.
Lua filters only change Pandoc's AST representation of your document, i.e. before it is then converted to LaTeX. A Raw block filter will not act on Pandoc's LaTeX output, but only on Raw LaTeX blocks that are in the markdown itself.

So, my idea was to write a custom Lua **writer**. The writer would use `pandoc.write` to generate Pandoc's own LaTeX output (body only) and modify it with regular expressions or Lua patterns. 

Unfortunately, I have no idea how to do this... ;/
See my updated script at https://github.com/produnis/minimal_examples/tree/longlist

**Comment by produnis** [comment, marked as answer]
I came up with a new dirty hack!

I've implemented a custom workaround to simulate a callout box in my Quarto project. Instead of using the built-in `:::{callout}` syntax, I created a new `:::{.longlistcallout}` class and added a Lua filter to style it for LaTeX and HTML output.

This looks just nice and works for me.

See https://github.com/produnis/minimal_examples/tree/longlist for the complete minimal example and my lua script.

**Reply by Steinthal** [reply]
FYI: Your minimal example works for me just with ```callout-note``` for PDF generation.  I'm currently using Quarto 1.6.32 (RC for 1.6):


![CleanShot 2024-11-02 at 21 44 57@2x](https://github.com/user-attachments/assets/4f4ad4be-a5b9-40dc-a7cf-99d66910bf1d)

**Reply by produnis** [reply]
Thx for letting me know, good to hear it is fixed in future Version

**Reply by mcanouil** [reply]
FYI, 1.6.32 is a Release Candidate and is the default download on Quarto site.
